controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -0,0 +1,513 @@
+package externalworkload
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	ewv1alpha1 "github.com/linkerd/linkerd2/controller/gen/apis/externalworkload/v1alpha1"
+	"github.com/linkerd/linkerd2/controller/k8s"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	logging "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	// Specifies capacity for updates buffer
+	updateQueueCapacity = 400
+
+	// Name of the lease resource the controller will use
+	leaseName = "linkerd-destination-endpoint-write"
+
+	// Duration of the lease
+	leaseDuration = 30 * time.Second
+
+	// Deadline for the leader to refresh its lease. Core controllers have a
+	// deadline of 10 seconds.
+	leaseRenewDeadline = 10 * time.Second
+
+	// Duration a leader elector should wait in between action re-tries.
+	// Core controllers have a value of 2 seconds.
+	leaseRetryPeriod = 2 * time.Second
+)
+
+type (
+	updateQueue struct {
+		queue chan queueUpdate
+		queueMetrics
+	}
+
+	queueMetrics struct {
+		queueLength  prometheus.Gauge
+		queueUpdates prometheus.Counter
+		queueDrops   prometheus.Counter
+		queueLatency prometheus.Histogram
+	}
+
+	queueUpdate struct {
+		item string
+
+		enqueTime time.Time
+	}
+)
+
+var (
+	queueLengthGauge = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "external_endpoints_controller_queue_length",
+		Help: "Total number of updates currently waiting to be processed",
+	})
+
+	queueUpdateCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "external_endpoints_controller_queue_updates",
+		Help: "Total number of updates that entered the queue",
+	})
+
+	queueDroppedCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "external_endpoints_controller_queue_dropped",
+		Help: "Total number of updates dropped due to a backed-up queue",
+	})
+
+	queueLatencyHist = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name:    "external_endpoints_controller_queue_latency_seconds",
+		Help:    "Distribution of durations that updates have spent in the queue",
+		Buckets: []float64{.005, .01, .05, .1, .5, 1, 3, 10},
+	})
+)
+
+// EndpointsController reconciles service memberships for ExternalWorkload resources
+// by writing EndpointSlice objects for Services that select one or more
+// external endpoints.
+type EndpointsController struct {
+	k8sAPI   *k8s.API
+	log      *logging.Entry
+	updates  updateQueue
+	stop     chan struct{}
+	isLeader bool
```

Comment:
making this an atomic bool is probably more straightforward and less error prone than protecting it with a mutex.

Reply:
great suggestion to simplify this 👍🏻